Add support for partitioning with PetscSF

stefanozampini opened the pull request:
This implements a backend for partitioning based on the `PetscSF` object https://arxiv.org/abs/2102.13018

It comes in two different flavors:
- `PETScWrappers::CommunicationPattern` is an implementation of `Utilities::MPI::NoncontiguousPartitioner`
- `PETScWrappers::Partitioner` is an implementation of `Utilities::MPI::Partitioner`

There are a few differences wrt the dealII classes:
- API based: PETSc does not need explicit specification of the communication channels, and the user passing MPI_Requests and temporary storage
- the `import_from_ghosted` does not explicitly zero the ghost data after communication is performed and its signature is thus `ArrayView<const ...>` (I don't see what is the point of zeroing the ghost data within the internal API when users can do that themselves if needed)

Support for device memory can be easily added by just adding an extra template argument on the memory type of the `ArrayView` arguments, since PetscSF already support device memory pointers

Tests are copy-pasted from already existing tests.

stefanozampini commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+    AssertPETSc(PetscSFDestroy(&sf));
+  }
+
+  const MPI_Comm &
```
Here you have an idea of how we need to implement `get_mpi_communicator()` if we keep the reference in the return argument. We cannot use the static variable inside the function since that will get set only once and never updated if the communicator changes.

Removing the reference for those methods for vectors, matrices etc will imply changing the methods of all the other implementations in deal.II for consistency.

stefanozampini commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+    AssertPETSc(PetscSFDestroy(&sf));
+  }
+
+  const MPI_Comm &
+  CommunicationPattern::get_mpi_communicator() const
+  {
+    this->returncomm = PetscObjectComm(reinterpret_cast<PetscObject>(sf));
+    return this->returncomm;
+  }
```
Here you have an idea of how we need to implement get_mpi_communicator() if we keep the reference in the return argument. We cannot use the static variable inside the function since that will get set only once and never be updated if the communicator changes.

Removing the reference for those methods for vectors, matrices, etc will imply changing the methods of all the other implementations in deal.II for consistency.

bangerth replied:
Yes, we should remove the references indeed

stefanozampini replied:
I will change this to return MPI_Comm and then modify all other methods in another PR

bangerth commented:
Ah, the famous star-forest finally makes an appearance :-)

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+
+DEAL_II_NAMESPACE_OPEN
+// Shorthand notation for PETSc error codes.
+#  define AssertPETSc(code)                              \
+    do                                                   \
+      {                                                  \
+        PetscErrorCode __ierr = (code);                  \
+        AssertThrow(__ierr == 0, ExcPETScError(__ierr)); \
+      }                                                  \
+    while (0)
```
This might be something to move into `exceptions.h` eventually.

bangerth replied:
One wished that PETSc was written in C++ and indicated error conditions by way of an exception. Then we would not have to obfuscate the code for every single PETSc call...

stefanozampini replied:
Yeah, I know. I'd rather leave this change to a quite big PR with changes to all the PETScWrappers

bangerth replied:
Yes, absolutely!

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+    want.add_indices(ghost_indices);
+    want.compress();
+
+    const PetscInt *idxs;
+    PetscInt        n;
+    IS              is = want.make_petsc_is(communicator);
+    AssertPETSc(ISGetLocalSize(is, &n));
```
I don't understand this code. The `IS` object so created just gets a bunch of indices on every process, but no indication which of these are locally owned. How does it determine which ones of these are local?

stefanozampini replied:
`PetscSFSetGraphLayout(sf, layout, n, nullptr, PETSC_OWN_POINTER, idxs))` does this for you.

stefanozampini replied:
Using `IS` is superfluous here (I could have used a `std::vector<PetscInt>` instead), it is only to test the new functionality in `IndexSet`

bangerth replied:
Still, I don't know what is happening in the three lines 66-68. You create an index set from a bunch of indices, and then `ISGetLocalSize` returns how many of these are stored on the current process: https://petsc.org/main/docs/manualpages/IS/ISGetLocalSize/ How does it know? Does `make_petsc_is()` internally partition indices into locally owned and ghost ones?

stefanozampini replied:
the local size corresponds to the number of indices stored in the array, independent of their actual distribution.
`IS` does not know anything about global indices distribution, it is merely a class around an array of indices.

stefanozampini replied:
So the indices returned are passed as "global indices I would like to receive" to the PetscSFSetGraphLayout call

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+  }
+
+  void
+  CommunicationPattern::reinit(const IndexSet &vector_space_vector_index_set,
+                               const IndexSet &read_write_vector_index_set,
```
These argument names are not particularly good (but not your fault). If you were so inclined, would you use the names of #14740 ?

stefanozampini replied:
Will do

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+    bool     has_invalid = false;
+    for (const auto i : indices_has)
+      {
+        if (i != numbers::invalid_dof_index)
```
How can this happen? An `IndexSet` should never store such an index.

stefanozampini replied:
This `reinit` accepts `std::vector`, not `IndexSet`. It is used e.g. here  https://github.com/dealii/dealii/pull/14736/files/3dd5da076e273db6f1f48f24a5aad4effb9a0f0d#diff-c8636b9e39678217cab6f2d9f405735b74b68706d67096b83397caa07ff7a0e9R77

bangerth replied:
Oh oh, of course!

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+
+    loc         = 0;
+    has_invalid = false;
+    for (const auto i : indices_want)
+      {
+        if (i != numbers::invalid_dof_index)
```
same question here

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+    // First create an SF where leaves are inidx (at location inloc)
+    // and roots are unique indices in contiguous way
+    // Code adapted from MatZeroRowsMapLocal_Private in PETSc
+    PetscInt n  = static_cast<PetscInt>(inidx.size()), nl;
```
I read this wrong the first time around, believing that you're using the comma operator here. Let's do this instead:
```suggestion
    const PetscInt n  = static_cast<PetscInt>(inidx.size());
    PetscInt nl;
```

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+
+    const PetscInt *ranges;
+    AssertPETSc(PetscLayoutGetRanges(layout, &ranges));
+#  if 0
+    PetscPrintf(PETSC_COMM_WORLD, "INIDX\n");
+    PetscIntView(inidx.size(), inidx.data(), PETSC_VIEWER_STDOUT_WORLD);
+    PetscPrintf(PETSC_COMM_WORLD, "INLOC\n");
+    PetscIntView(inloc.size(),
+                 inloc.size() > 0 ? inloc.data() : nullptr,
+                 PETSC_VIEWER_STDOUT_WORLD);
+    PetscPrintf(PETSC_COMM_WORLD, "OUTIDX\n");
+    PetscIntView(outidx.size(), outidx.data(), PETSC_VIEWER_STDOUT_WORLD);
+    PetscPrintf(PETSC_COMM_WORLD, "OUTLOC\n");
+    PetscIntView(outloc.size(),
+                 outloc.size() > 0 ? outloc.data() : nullptr,
+                 PETSC_VIEWER_STDOUT_WORLD);
+    if (PetscGlobalRank == 0)
+      {
+        PetscPrintf(PETSC_COMM_WORLD, "RANGES\n");
+        PetscMPIInt size;
+        MPI_Comm_size(communicator, &size);
+        PetscIntView(size + 1, ranges, PETSC_VIEWER_STDOUT_SELF);
+      }
+#  endif
```
Let's remove this, I assume that you've tested the code.

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+      PETSC_COPY_VALUES,
+      const_cast<PetscInt *>(n > 0 ? outidx.data() : nullptr)));
+    AssertPETSc(PetscSFSetUp(sf2));
+    // AssertPETSc(PetscSFView(sf2, nullptr));
```
If you don't need it any more, remove this line -- same four lines down

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+    return this->returncomm;
+  }
+
+  template <typename Number>
+  void
+  CommunicationPattern::export_to_ghosted_array_start(
+    const ArrayView<const Number> &src,
+    const ArrayView<Number> &      dst) const
+  {
+    auto datatype = Utilities::MPI::mpi_type_id_for_type<Number>;
+
+#  if DEAL_II_PETSC_VERSION_LT(3, 15, 0)
+    AssertPETSc(PetscSFBcastBegin(sf, datatype, src.data(), dst.data()));
+#  else
+    AssertPETSc(
+      PetscSFBcastBegin(sf, datatype, src.data(), dst.data(), MPI_REPLACE));
+#  endif
+  }
+
+  template <typename Number>
+  void
+  CommunicationPattern::export_to_ghosted_array_finish(
+    const ArrayView<const Number> &src,
+    const ArrayView<Number> &      dst) const
+  {
+    auto datatype = Utilities::MPI::mpi_type_id_for_type<Number>;
+
+#  if DEAL_II_PETSC_VERSION_LT(3, 15, 0)
+    AssertPETSc(PetscSFBcastEnd(sf, datatype, src.data(), dst.data()));
+#  else
+    AssertPETSc(
+      PetscSFBcastEnd(sf, datatype, src.data(), dst.data(), MPI_REPLACE));
+#  endif
+  }
```
One day I'm going to replace all of these `_begin/start()` and `_end/finish()` with the facilities of the `MPI::Future` class in #14039 :-)

bangerth commented on source/lac/petsc_communication_pattern.cc:
```diff
@@ -0,0 +1,516 @@
+  {
+    export_to_ghosted_array_start(src, dst);
+    export_to_ghosted_array_finish(src, dst);
+  }
+
+  template <typename Number>
```
Total aside: We generally have three empty lines between functions. No need to change anything here, eventually someone will clean it up, but if you want to keep it in mind for future patches :-)

bangerth reviewed (commented):
Here are some comment.

I don't know my way around the partitioners very well. It would be nice if @kronbichler or @peterrum would take another look at this, but I think most of it is pretty clear at least as far as the interface is concerned. Thanks for adding all of these tests!

stefanozampini commented on include/deal.II/lac/petsc_communication_pattern.h:
```diff
@@ -0,0 +1,402 @@
+  /**
+   * CommunicationPattern implementation based on the PetscSF object.
+   *
+   * For additional information, see the paper
+   *
+   *    The PetscSF Scalable Communication Layer
+   *
+   * by J. Zhang and J. Brown and S. Balay and J. Faibussowitsch and M.
+   * Knepley and O. Marin and R. Mills and T. Munson and B. F. Smith and S.
+   * Zampini
+   *
+   * IEEE Transactions on Parallel and Distributed Systems 33 (4),
+   *
+   * doi = 10.1109/TPDS.2021.3084070
```
Will do

stefanozampini commented on include/deal.II/lac/petsc_communication_pattern.h:
```diff
@@ -0,0 +1,402 @@
+     */
+    virtual ~CommunicationPattern() override;
+
+    /**
+     * Reinitialize the communication pattern. The first argument @p
+     * vector_space_vector_index_set is the index set associated to a
+     * VectorSpaceVector object. The second argument @p
+     * read_write_vector_index_set is the index set associated to a
+     * ReadWriteVector object.
+     *
+     * The export_to_ghost_array will populate an array with the values
+     * associated to the @p read_write_vector_index_set only.
+     */
+    virtual void
+    reinit(const IndexSet &vector_space_vector_index_set,
+           const IndexSet &read_write_vector_index_set,
+           const MPI_Comm &communicator) override;
+
```
Do I have to do anything special to "let doxygen inherit it from the base class"?

bangerth replied:
Nope, we configured doxygen so that it automatically inherits the documentation of undocumented virtual functions from the base class.

bangerth commented:
Let me know once you've got it in a state where you want me to take another look!

stefanozampini commented:
@bangerth I think I have addressed your comments. Please take a look.
I have decided to implement get_mpi_communicator with a static variable for the moment, since otherwise I would have to change all the implementations of the partitioners. I will do all the get_mpi_communicator changes in a single PR

stefanozampini commented:
/rebuild

bangerth reviewed (approved):
My apologies for not getting to this faster. This looks good now, many thanks!

Out of curiosity, why did you implement this instead of just using the existing classes that do the same thing? Are your classes substantially faster?
